gui: Fix race in WalletModel::pollBalanceChanged

ryanofsky opened the pull request:
Poll function was wrongly setting cached height to the current chain height instead of the chain height at the time of polling.

This bug could cause balances to appear out of date, and was first introduced https://github.com/bitcoin/bitcoin/pull/10244/commits/a0704a8996bb950ae3c4d5b5a30e9dfe34cde1d3#diff-2e3836af182cfb375329c3463ffd91f8L117. Before that commit, there wasn't a problem because cs_main was held during the poll update.

Currently, the problem should be rare. But if 8937d99ce81a27ae5e1012a28323c0e26d89c50b from #17954 were merged, the problem would get worse, because the wrong cachedNumBlocks value would be set if the wallet was polled in the interval between a block being connected and it processing the BlockConnected notification.

MarcoFalke also points out that a0704a8996b could lead to GUI hangs as well, because previously the pollBalanceChanged method, which runs on the GUI thread, would only make a nonblocking TRY_LOCK(cs_main) call, but after could make blocking LOCK(cs_main) calls, potentially locking up the GUI.

Thanks to John Newbery for finding this bug this while reviewing https://github.com/bitcoin/bitcoin/pull/17954.

ryanofsky committed:
gui: Fix race in WalletModel::pollBalanceChanged

Poll function was wrongly setting cached height to the current chain height
instead of the chain height at the time of polling.

This bug could cause balances to appear out of date, and was first introduced
https://github.com/bitcoin/bitcoin/pull/10244/commits/a0704a8996bb950ae3c4d5b5a30e9dfe34cde1d3#r378452145
Before that commit, there wasn't a problem because cs_main was held during the
poll update.

Currently, the problem should be rare. But if
8937d99ce81a27ae5e1012a28323c0e26d89c50b from #17954 were merged, the problem
would get worse, because the wrong cachedNumBlocks value would be set if the
wallet was polled in the interval between a block being connected and it
processing the BlockConnected notification.

MarcoFalke <falke.marco@gmail.com> also points out that a0704a8996b could lead
to GUI hangs as well, because previously the pollBalanceChanged method, which
runs on the GUI thread, would only make a nonblocking TRY_LOCK(cs_main) call,
but after could make blocking LOCK(cs_main) calls, potentially locking up the
GUI.

Thanks to John Newbery <john@johnnewbery.com> for finding this bug this while
reviewing https://github.com/bitcoin/bitcoin/pull/17954.

MarcoFalke commented:
This might be another reason that users are seeing issues like #17112 . We might want to consider this for backport

luke-jr commented:
utACK

jonasschnelli commented:
Good catch and thanks for the fix.
Agree with @MarcoFalke that this should be backported.

utACK 37d27bc070553505568cda900d22b5a3600fb54c

ryanofsky commented:
Updated 37d27bc070553505568cda900d22b5a3600fb54c -> bf36a3ccc212ad4d7c5cb8f26d7a22e279fe3cec ([`pr/pollbug.1`](https://github.com/ryanofsky/bitcoin/commits/pr/pollbug.1) -> [`pr/pollbug.2`](https://github.com/ryanofsky/bitcoin/commits/pr/pollbug.2), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/pollbug.1...pr/pollbug.2)). No code changes, just updated the commit description to mention how this could be related to GUI hangs

Empact commented:
utACK https://github.com/bitcoin/bitcoin/pull/18123/commits/37d27bc070553505568cda900d22b5a3600fb54c

jonasschnelli commented:
utACK bf36a3c

promag commented:
ACK bf36a3ccc212ad4d7c5cb8f26d7a22e279fe3cec.

> This might be another reason that users are seeing issues like #17112 . We might want to consider this for backport

@MarcoFalke `tryGetBalance` just avoids waiting for the lock, is doesn't save you the time spent in `CWallet::GetBalance`, so I don't see how this could make it better. See #16874.

ryanofsky commented:
> `tryGetBalance` just avoids waiting for the lock, is doesn't save you the time spent in `CWallet::GetBalance`, so I don't see how this could make it better

@promag The case where this could alleviate GUI hangs is the case where `tryGetBalance` *succeeds* in getting the locks and balance without blocking, but then before the one of the two `m_node.getNumBlocks()` something else acquires `cs_main`, and the GUI thread blocks in the `m_node.getNumBlocks()` call

promag commented:
@ryanofsky ah! thanks for explaining. Nice catch @jnewbery.

fanquake commented:
Being backported in 18218.
